### ISSUE TITLE
firmware name/version is cleared when restarting tedge-agent

### ISSUE THREAD

**Issue by reubenmiller** [thin-edge/thin-edge.io#2497]
**Describe the bug**

The current firmware name/version/url (set after processing a successful firmware_update operation) is cleared in the Cumulocity Cloud when the tedge-agent service is restarted when using the new firmware_update workflow.

The error is caused because the `te/<identity>/cmd/firmware_update` topic is being used for different purposes

* Communicate that the tedge-agent can handle firmware update operations
* Communicate the current firmware name/version

Whilst processing a firmware_update command, the name/version from the `te/<identity>/cmd/firmware_update/<cmd_id>` is read and the latest version

**To Reproduce**

Note: In order to test this you need to familiar with a very new feature which is only available from the main branch (and being actively developed).

1. Trigger a firmware_update operation
2. Wait for the operation to be completed
3. Check the device's firmware name/version in Cumulocity IoT UI
4. Restart the tedge-agent
5. View that the device's firmware name/version has now be cleared in Cumulocity IoT

**Expected behavior**

Restart a service should not clear the existing firmware name/version information.

The exact solution is unclear, however it seems that the firmware name/version might need to be decoupled from the command topic (but not 100% certain)

**Screenshots**

**Environment (please complete the following information):**

|Property|Value|
|--------|-----|
|OS [incl. version]|`Debian GNU/Linux 12 (bookworm)`|
|Hardware [incl. revision]|`Raspberry Pi 5 Model B Rev 1.0`|
|System-Architecture|`Linux rpi5-d83add9f145a 6.1.0-rpi4-rpi-2712 #1 SMP PREEMPT Debian 1:6.1.54-1+rpt2 (2023-10-05) aarch64 GNU/Linux`|
|thin-edge.io version|`tedge 0.13.2~119+g6266033`|

**Additional context**

**Comment by rina23q** [comment]
## Solution proposal
This issue can be fixed by 

- Remove contents of `FirmwareMetadata`
- User should use the device twin feature to declare firmware info
- Upon a successful firmware_update command, the mapper publishes the device twin message to update the content

### Remove contents of FirmwareMetadata
Keeping firmware info in `Commands::FirmwareMetadata` was a design mistake. `FirmwareMetadata` must be used only to declare device's capability. So, change the expected payload of [Command Metadata spec for firmware](https://github.com/thin-edge/thin-edge.io/issues/2453) to:

Topic: `te/<identifier>/cmd/firmware_update`
**Payload: can be `{}` empty JSON. If user provides more fields, mapper ignores them.**

### Device twin feature to declare firmware info
Firmware info is a data for inventory. Thus, using the device twin feature is appropriate to record the data in Thin Edge.

#### Thin Edge (input)
Topic: `te/<identifier>///twin/firmware`
Payload: No field name restriction. The `name`, `version`, and `url` fields are recommended to c8y users.
```json
{
  "name": "foo",
  "version": "1.0",
  "url": "https://example.com/20231211"
}
```

#### C8y (mapper output)
Topic: `c8y/inventory/managedObjects/update/<external-id>`
Payload: **The key `firmware` must renamed to `c8y_Firmware`**. Fields under `c8y_Firmware` are just converted from the original payload of the device twin topic. 
```json
{
  "c8y_Firmware": {
    "name": "foo",
    "version": "1.0",
    "url": "https://example.com/20231211"
  }
}
```

### On a successful firmware_update command, publish a device twin message
When c8y-mapper receives a successful message for firmware_update, the mapper should update the device twin. Take the data (`name`, `url`, and `version`) from the received command message, and publish it onto the `te/<identifier>///twin/firmware` topic.

## Alternatives
Use a new command channel `FirmwareInfo` like the relationship between `SoftwareUpdate` and `SoftwareList`. However, it's not trivial as firmware is always a single entry.

**Comment by rina23q** [comment]
Implemented by #2531 

### firmware_update metadata is now used only to report supported operation
If user publishes any JSON (recommended empty JSON `{}`) messages onto the firmware metadata topic `te/device/<...>///cmd/firmware_update`, c8y-mapper creates a supported operation file to `/etc/tedge/operations/c8y/c8y_Firmware`, then user should be able to see "Firmware" tab in their c8y tenant. However, user cannot find installed firmware info at this point in c8y.

### twin topic for firmware
If user wants to see the current firmware info, user needs to publish a JSON message to the device twin topic `te/device/<...>///twin/firmware`. The fields `url`, `name`, and `version` are optional, but recommended. Also, the `version` and `name` fields will be rendered in c8y tenant.

```json
{
   "url": "http://some.url",
   "name": "foo-bar-firmware",
   "version": "1.02-20230101"
}
```

This info will be converted as `c8y_Firmware` as described [here](https://github.com/thin-edge/thin-edge.io/issues/2497#issuecomment-1849946757).

### After a successful firmware operation
After a firmware operation is marked successful, c8y-mapper publishes the newly installed firmware info to the device twin topic as described above how the twin topic works for firmware. Previously, it was onto firmware metadata topic.

**Comment by gligorisaev** [comment]
QA has thoroughly checked the bug and here are the results:

- [ ] Test for ticket exists in the test suite.
- [x] QA has tested the bug and it's not reproducable anymore
